File: OneSignalSDK/onesignal/src/main/java/com/onesignal/NotificationOpenedProcessor.java
/**
 * Modified MIT License
 *
 * Copyright 2017 OneSignal
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * 2. All copies of substantial portions of the Software may only be used in connection
 * with services provided by OneSignal.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.onesignal;

import android.app.Activity;
import android.app.NotificationManager;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationManagerCompat;

import com.onesignal.OneSignalDbContract.NotificationTable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import static com.onesignal.GenerateNotification.BUNDLE_KEY_ANDROID_NOTIFICATION_ID;
import static com.onesignal.GenerateNotification.BUNDLE_KEY_ONESIGNAL_DATA;

// Process both notifications opens and dismisses.
class NotificationOpenedProcessor {

   private static final String TAG = NotificationOpenedProcessor.class.getCanonicalName();

   static void processFromContext(Context context, Intent intent) {
      if (!isOneSignalIntent(intent))
         return;

      if (context != null)
         OneSignal.initWithContext(context.getApplicationContext());

      handleDismissFromActionButtonPress(context, intent);

      processIntent(context, intent);
   }

   // Was Bundle created from our SDK? Prevents external Intents
   // TODO: Could most likely be simplified checking if BUNDLE_KEY_ONESIGNAL_DATA is present
   private static boolean isOneSignalIntent(Intent intent) {
      return intent.hasExtra(BUNDLE_KEY_ONESIGNAL_DATA) || intent.hasExtra("summary") || intent.hasExtra(BUNDLE_KEY_ANDROID_NOTIFICATION_ID);
   }

   private static void handleDismissFromActionButtonPress(Context context, Intent intent) {
      // Pressed an action button, need to clear the notification and close the notification area manually.
      if (intent.getBooleanExtra("action_button", false)) {
         NotificationManagerCompat.from(context).cancel(intent.getIntExtra(BUNDLE_KEY_ANDROID_NOTIFICATION_ID, 0));
         context.sendBroadcast(new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS));
      }
   }

   static void processIntent(Context context, Intent intent) {
      OneSignalDbHelper dbHelper = OneSignalDbHelper.getInstance(context);
      String summaryGroup = intent.getStringExtra("summary");

      boolean dismissed = intent.getBooleanExtra("dismissed", false);

      OSNotificationIntentExtras intentExtras = null;
      if (!dismissed) {
         intentExtras = processToOpenIntent(context, intent, dbHelper, summaryGroup);

         if (intentExtras == null)
            return;
      }

      markNotificationsConsumed(context, intent, dbHelper, dismissed);

      // Notification is not a summary type but a single notification part of a group.
      if (summaryGroup == null) {
         String group = intent.getStringExtra("grp");
         if (group != null)
            NotificationSummaryManager.updateSummaryNotificationAfterChildRemoved(context, dbHelper, group, dismissed);
      }

      OneSignal.onesignalLog(OneSignal.LOG_LEVEL.DEBUG, "processIntent from context: " + context + " and intent: " + intent);
      if (intent.getExtras() != null)
         OneSignal.onesignalLog(OneSignal.LOG_LEVEL.DEBUG, "processIntent intent extras: " + intent.getExtras().toString());

      if (!dismissed) {
         if (!(context instanceof Activity))
            OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "NotificationOpenedProcessor processIntent from an non Activity context: " + context);
         else OneSignal.handleNotificationOpen((Activity) context, intentExtras.getDataArray(),
                 false, OSNotificationFormatHelper.getOSNotificationIdFromJson(intentExtras.getJsonData()));
      }
   }

   static OSNotificationIntentExtras processToOpenIntent(Context context, Intent intent, OneSignalDbHelper dbHelper, String summaryGroup) {
      JSONArray dataArray = null;
      JSONObject jsonData = null;
      try {
         jsonData = new JSONObject(intent.getStringExtra(BUNDLE_KEY_ONESIGNAL_DATA));

         if (!(context instanceof Activity))
            OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "NotificationOpenedProcessor processIntent from an non Activity context: " + context);
         else if (handleIAMPreviewOpen((Activity) context, jsonData))
            return null;

         jsonData.put(BUNDLE_KEY_ANDROID_NOTIFICATION_ID, intent.getIntExtra(BUNDLE_KEY_ANDROID_NOTIFICATION_ID, 0));
         intent.putExtra(BUNDLE_KEY_ONESIGNAL_DATA, jsonData.toString());
         dataArray = NotificationBundleProcessor.newJsonArray(new JSONObject(intent.getStringExtra(BUNDLE_KEY_ONESIGNAL_DATA)));
      } catch (JSONException e) {
         e.printStackTrace();
      }

      // We just opened a summary notification.
      if (summaryGroup != null)
         addChildNotifications(dataArray, summaryGroup, dbHelper);

      return new OSNotificationIntentExtras(dataArray, jsonData);
   }

   static boolean handleIAMPreviewOpen(@NonNull Activity context, @NonNull JSONObject jsonData) {
      String previewUUID = OSInAppMessagePreviewHandler.inAppPreviewPushUUID(jsonData);
      if (previewUUID == null)
         return false;

      OneSignal.getInAppMessageController().displayPreviewMessage(previewUUID);
      return true;
   }

   private static void addChildNotifications(JSONArray dataArray, String summaryGroup, OneSignalDbHelper writableDb) {
      String[] retColumn = { NotificationTable.COLUMN_NAME_FULL_DATA };
      String[] whereArgs = { summaryGroup };

      Cursor cursor = writableDb.query(
            NotificationTable.TABLE_NAME,
            retColumn,
            NotificationTable.COLUMN_NAME_GROUP_ID + " = ? AND " +   // Where String
                  NotificationTable.COLUMN_NAME_DISMISSED + " = 0 AND " +
                  NotificationTable.COLUMN_NAME_OPENED + " = 0 AND " +
                  NotificationTable.COLUMN_NAME_IS_SUMMARY + " = 0",
            whereArgs,
            null, null, null);

      if (cursor.getCount() > 1) {
         cursor.moveToFirst();
         do {
            try {
               String jsonStr = cursor.getString(cursor.getColumnIndex(NotificationTable.COLUMN_NAME_FULL_DATA));
               dataArray.put(new JSONObject(jsonStr));
            } catch (JSONException e) {
               OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Could not parse JSON of sub notification in group: " + summaryGroup);
            }
         } while (cursor.moveToNext());
      }

      cursor.close();
   }

   private static void markNotificationsConsumed(Context context, Intent intent, OneSignalDbHelper writableDb, boolean dismissed) {
      String summaryGroup = intent.getStringExtra("summary");
      String whereStr;
      String[] whereArgs = null;

      if (summaryGroup != null) {
         boolean isGroupless = summaryGroup.equals(OneSignalNotificationManager.getGrouplessSummaryKey());
         if (isGroupless)
            whereStr = NotificationTable.COLUMN_NAME_GROUP_ID + " IS NULL";
         else {
            whereStr = NotificationTable.COLUMN_NAME_GROUP_ID + " = ?";
            whereArgs = new String[]{ summaryGroup };
         }

         if (!dismissed) {
            // Make sure when a notification is not being dismissed it is handled through the dashboard setting
            boolean shouldDismissAll = OneSignal.getClearGroupSummaryClick();
            if (!shouldDismissAll) {
               /* If the open event shouldn't clear all summary notifications then the SQL query
                * will look for the most recent notification instead of all grouped notifications */
               String mostRecentId = String.valueOf(OneSignalNotificationManager.getMostRecentNotifIdFromGroup(writableDb, summaryGroup, isGroupless));
               whereStr += " AND " + NotificationTable.COLUMN_NAME_ANDROID_NOTIFICATION_ID + " = ?";
               whereArgs = isGroupless ?
                       new String[]{ mostRecentId } :
                       new String[]{ summaryGroup, mostRecentId };
            }
         }
      } else
         whereStr = NotificationTable.COLUMN_NAME_ANDROID_NOTIFICATION_ID + " = " + intent.getIntExtra(BUNDLE_KEY_ANDROID_NOTIFICATION_ID, 0);

      clearStatusBarNotifications(context, writableDb, summaryGroup);
      writableDb.update(NotificationTable.TABLE_NAME, newContentValuesWithConsumed(intent), whereStr, whereArgs);
      BadgeCountUpdater.update(writableDb, context);
   }

   /**
    * Handles clearing the status bar notifications when opened
    */
   private static void clearStatusBarNotifications(Context context, OneSignalDbHelper writableDb, String summaryGroup) {
      // Handling for clearing the notification when opened
      if (summaryGroup != null)
         NotificationSummaryManager.clearNotificationOnSummaryClick(context, writableDb, summaryGroup);
      else {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            // The summary group is null, represents the last notification in the groupless group
            // Check that no more groupless notifications exist in the group and cancel the group
            int grouplessCount = OneSignalNotificationManager.getGrouplessNotifsCount(context);
            if (grouplessCount < 1) {
               int groupId = OneSignalNotificationManager.getGrouplessSummaryId();
               NotificationManager notificationManager = OneSignalNotificationManager.getNotificationManager(context);
               notificationManager.cancel(groupId);
            }
         }
      }
   }

   private static ContentValues newContentValuesWithConsumed(Intent intent) {
      ContentValues values = new ContentValues();

      boolean dismissed = intent.getBooleanExtra("dismissed", false);

      if (dismissed)
         values.put(NotificationTable.COLUMN_NAME_DISMISSED, 1);
      else
         values.put(NotificationTable.COLUMN_NAME_OPENED, 1);

      return values;
   }

}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/notification/internal/actions/NotificationOpenedProcessor.kt
/**
 * Modified MIT License
 *
 * Copyright 2017 OneSignal
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * 2. All copies of substantial portions of the Software may only be used in connection
 * with services provided by OneSignal.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.onesignal.onesignal.notification.internal.actions

import android.content.Intent
import android.app.Activity
import org.json.JSONException
import android.content.ContentValues
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationManagerCompat
import com.onesignal.onesignal.core.internal.common.JSONUtils
import com.onesignal.onesignal.core.internal.database.IDatabaseProvider
import com.onesignal.onesignal.core.internal.database.impl.OneSignalDbContract
import com.onesignal.onesignal.notification.internal.NotificationHelper
import com.onesignal.onesignal.notification.internal.data.INotificationDataController
import com.onesignal.onesignal.notification.internal.data.NotificationSummaryManager
import com.onesignal.onesignal.core.internal.params.IParamsService
import com.onesignal.onesignal.core.internal.logging.Logging
import com.onesignal.onesignal.notification.internal.badges.BadgeCountUpdater
import com.onesignal.onesignal.notification.internal.generation.GenerateNotification
import org.json.JSONArray
import org.json.JSONObject

// Process both notifications opens and dismisses.
internal class NotificationOpenedProcessor(
    private val _databaseProvider: IDatabaseProvider,
    private val _summaryManager: NotificationSummaryManager,
    private val _dataController: INotificationDataController,
    private val _paramsService: IParamsService,
    private val _badgeUpdater: BadgeCountUpdater
) {
    private val TAG = NotificationOpenedProcessor::class.java.canonicalName
    suspend fun processFromContext(context: Context, intent: Intent) {
        if (!isOneSignalIntent(intent))
            return

        handleDismissFromActionButtonPress(context, intent)
        processIntent(context, intent)
    }

    // Was Bundle created from our SDK? Prevents external Intents
    // TODO: Could most likely be simplified checking if BUNDLE_KEY_ONESIGNAL_DATA is present
    private fun isOneSignalIntent(intent: Intent): Boolean {
        return intent.hasExtra(GenerateNotification.BUNDLE_KEY_ONESIGNAL_DATA) || intent.hasExtra("summary") || intent.hasExtra(
            GenerateNotification.BUNDLE_KEY_ANDROID_NOTIFICATION_ID
        )
    }

    private fun handleDismissFromActionButtonPress(context: Context?, intent: Intent) {
        // Pressed an action button, need to clear the notification and close the notification area manually.
        if (intent.getBooleanExtra("action_button", false)) {
            NotificationManagerCompat.from(context!!).cancel(
                intent.getIntExtra(
                    GenerateNotification.BUNDLE_KEY_ANDROID_NOTIFICATION_ID,
                    0
                )
            )

            // Only close the notifications shade on Android versions where it is allowed, Android 11 and lower.
            // See https://developer.android.com/about/versions/12/behavior-changes-all#close-system-dialogs
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
                context.sendBroadcast(Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS))
            }
        }
    }

    suspend fun processIntent(context: Context, intent: Intent) {
        val summaryGroup = intent.getStringExtra("summary")
        val dismissed = intent.getBooleanExtra("dismissed", false)
        var intentExtras: NotificationIntentExtras? = null
        if (!dismissed) {
            intentExtras = processToOpenIntent(context, intent, summaryGroup)
            if (intentExtras == null) return
        }
        markNotificationsConsumed(context, intent, dismissed)

        // Notification is not a summary type but a single notification part of a group.
        if (summaryGroup == null) {
            val group = intent.getStringExtra("grp")
            if (group != null)
                _summaryManager.updateSummaryNotificationAfterChildRemoved(context, group, dismissed)
        }
        Logging.debug("processIntent from context: $context and intent: $intent")
        if (intent.extras != null)
            Logging.debug("""processIntent intent extras: ${intent.extras.toString()}""")
        if (!dismissed) {
            if (context !is Activity)
                Logging.error("NotificationOpenedProcessor processIntent from an non Activity context: $context")
            else {
                // TODO: Implement
                //OneSignal.handleNotificationOpen(context as Activity?, intentExtras!!.dataArray, NotificationFormatHelper.getOSNotificationIdFromJson(intentExtras.jsonData))
            }
        }
    }

    fun processToOpenIntent(
        context: Context?,
        intent: Intent,
        summaryGroup: String?
    ): NotificationIntentExtras? {
        var dataArray: JSONArray? = null
        var jsonData: JSONObject? = null
        try {
            jsonData =
                JSONObject(intent.getStringExtra(GenerateNotification.BUNDLE_KEY_ONESIGNAL_DATA))

            if (context !is Activity)
                Logging.error("NotificationOpenedProcessor processIntent from an non Activity context: $context")
            // TODO: Implement
            //else if (OSInAppMessagePreviewHandler.notificationOpened((context as Activity?)!!, jsonData))
            //    return null

            jsonData.put(
                GenerateNotification.BUNDLE_KEY_ANDROID_NOTIFICATION_ID, intent.getIntExtra(
                    GenerateNotification.BUNDLE_KEY_ANDROID_NOTIFICATION_ID, 0
                )
            )
            intent.putExtra(GenerateNotification.BUNDLE_KEY_ONESIGNAL_DATA, jsonData.toString())
            dataArray = JSONUtils.wrapInJsonArray(
                JSONObject(
                    intent.getStringExtra(
                        GenerateNotification.BUNDLE_KEY_ONESIGNAL_DATA
                    )
                )
            )
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        // We just opened a summary notification.
        if (summaryGroup != null)
            addChildNotifications(dataArray, summaryGroup)

        return NotificationIntentExtras(dataArray, jsonData)
    }

    private fun addChildNotifications(dataArray: JSONArray?, summaryGroup: String) {

        val retColumn = arrayOf(OneSignalDbContract.NotificationTable.COLUMN_NAME_FULL_DATA)
        val whereArgs = arrayOf(summaryGroup)
        val cursor = _databaseProvider.get().query(
            OneSignalDbContract.NotificationTable.TABLE_NAME,
            retColumn,
            OneSignalDbContract.NotificationTable.COLUMN_NAME_GROUP_ID + " = ? AND " +  // Where String
                    OneSignalDbContract.NotificationTable.COLUMN_NAME_DISMISSED + " = 0 AND " +
                    OneSignalDbContract.NotificationTable.COLUMN_NAME_OPENED + " = 0 AND " +
                    OneSignalDbContract.NotificationTable.COLUMN_NAME_IS_SUMMARY + " = 0",
            whereArgs,
            null, null, null
        )
        if (cursor.count > 1) {
            cursor.moveToFirst()
            do {
                try {
                    val jsonStr: String =
                        cursor.getString(cursor.getColumnIndex(OneSignalDbContract.NotificationTable.COLUMN_NAME_FULL_DATA))
                    dataArray!!.put(JSONObject(jsonStr))
                } catch (e: JSONException) {
                    Logging.error("Could not parse JSON of sub notification in group: $summaryGroup")
                }
            } while (cursor.moveToNext())
        }
        cursor.close()
    }

    private suspend fun markNotificationsConsumed(
        context: Context,
        intent: Intent,
        dismissed: Boolean
    ) {
        val summaryGroup = intent.getStringExtra("summary")
        var whereStr: String
        var whereArgs: Array<String>? = null
        if (summaryGroup != null) {
            val isGroupless = summaryGroup == NotificationHelper.grouplessSummaryKey
            if (isGroupless)
                whereStr = OneSignalDbContract.NotificationTable.COLUMN_NAME_GROUP_ID + " IS NULL"
            else {
                whereStr = OneSignalDbContract.NotificationTable.COLUMN_NAME_GROUP_ID + " = ?"
                whereArgs = arrayOf(summaryGroup)
            }
            if (!dismissed) {
                // Make sure when a notification is not being dismissed it is handled through the dashboard setting
                if (!_paramsService.clearGroupOnSummaryClick) {
                    /* If the open event shouldn't clear all summary notifications then the SQL query
                * will look for the most recent notification instead of all grouped notifications */
                    val mostRecentId = _dataController.getMostRecentNotifIdFromGroup(summaryGroup, isGroupless).toString()
                    whereStr += " AND " + OneSignalDbContract.NotificationTable.COLUMN_NAME_ANDROID_NOTIFICATION_ID + " = ?"
                    whereArgs = if (isGroupless) arrayOf(mostRecentId) else arrayOf(
                        summaryGroup,
                        mostRecentId
                    )
                }
            }
        } else
            whereStr = OneSignalDbContract.NotificationTable.COLUMN_NAME_ANDROID_NOTIFICATION_ID + " = " + intent.getIntExtra(
                    GenerateNotification.BUNDLE_KEY_ANDROID_NOTIFICATION_ID, 0
                )

        clearStatusBarNotifications(context, summaryGroup)
        _databaseProvider.get().update(
            OneSignalDbContract.NotificationTable.TABLE_NAME,
            newContentValuesWithConsumed(intent),
            whereStr,
            whereArgs
        )

        _badgeUpdater.update(context)
    }

    /**
     * Handles clearing the status bar notifications when opened
     */
    private suspend fun clearStatusBarNotifications(context: Context, summaryGroup: String?) {
        // Handling for clearing the notification when opened
        if (summaryGroup != null)
            _summaryManager.clearNotificationOnSummaryClick(context, summaryGroup)
        else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                // The summary group is null, represents the last notification in the groupless group
                // Check that no more groupless notifications exist in the group and cancel the group
                val grouplessCount = NotificationHelper.getGrouplessNotifsCount(context)
                if (grouplessCount < 1) {
                    val groupId = NotificationHelper.grouplessSummaryId
                    val notificationManager =
                        NotificationHelper.getNotificationManager(context)
                    notificationManager.cancel(groupId)
                }
            }
        }
    }

    private fun newContentValuesWithConsumed(intent: Intent): ContentValues {
        val values = ContentValues()
        val dismissed = intent.getBooleanExtra("dismissed", false)
        if (dismissed) values.put(
            OneSignalDbContract.NotificationTable.COLUMN_NAME_DISMISSED,
            1
        ) else values.put(
            OneSignalDbContract.NotificationTable.COLUMN_NAME_OPENED, 1
        )
        return values
    }
}